### PR TITLE
Add support for running Kotlin/JS compiler

### PR DESCRIPTION
**Pull request by bnorm**

### CODE REVIEW

**Comment by bnorm** [comment]
The only thing I think is lacking from this PR is how to test the compiled Kotlin/JS when the compiler is done. With the JVM compiler you can just load the ClassLoader but that doesn't work for Kotlin/JS. I wonder if something like GraalVM can help here but I'm not sure: https://github.com/graalvm/graaljs

My personal use case was for https://github.com/bnorm/kotlin-react-function where I just wanted to have an embedded compiler I could add breakpoints to for debugging.

**Review by tschuchortdev — CHANGES_REQUESTED** [review]
Thanks for tackling this issue! At first glance it looks good but it needs more tests and the amount of code duplication makes this hard to maintain. I think the best way is probably to integrate the new functionality into the existing KotlinCompilation class and add an option to enable JS compilation. Otherwise a lot of stuff has to be factored out so it can be shared by the two KotlinCompilation classes, but that would be quite hard to do for all the logic. Long term I plan to rewrite it and make the whole thing more modular but I'm not ready to do that yet, given that it would break all API compatibility. 

> The only thing I think is lacking from this PR is how to test the compiled Kotlin/JS when the compiler is done [...] I wonder if something like GraalVM can help here but I'm not sure: https://github.com/graalvm/graaljs

I'm hesitant to add any functionality for this to the library right now if it's not straight forward. It's probably best to leave it in an "incomplete" state, let users figure out the best way to inspect generated JS files and wait for their feedback. But we can try GraalVM for internal library tests if you're willing to do that.

File: core/src/main/kotlin/com/tschuchort/compiletesting/KotlinJsCompilation.kt [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+
+    args.sourceMap = true
+    args.listPhases = true
+    args.moduleKind = "commonjs"
+    args.outputFile = File(outputDir, outputFileName).absolutePath
+    args.metaInfo = true
+    args.sourceMapBaseDirs = commonClasspaths().joinToString(separator = File.pathSeparator)
+    args.libraries = listOfNotNull(kotlinStdLibJsJar).joinToString(separator = ":")
+
+    args.irProduceKlibDir = irProduceKlibDir
```
[the remark this answers is not in the document]
Comment (bnorm) [reply core/src/main/kotlin/com/tschuchort/compiletesting/KotlinJsCompilation.kt:184]: These options are unique to the JS compiler. JVM compiler only has a `useIR` option. If we want to limit the number of options initially the important one is `irProduceJs` for testing IR based compiler plugins. This also brings up the question if the library should abstract these options or use the same naming as the underlying options.

File: core/src/main/kotlin/com/tschuchort/compiletesting/KotlinJsCompilation.kt [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+
+  /** Result of the compilation */
+  inner class Result(
+    /** The exit code of the compilation */
+    val exitCode: ExitCode,
+    /** Messages that were printed by the compilation */
+    val messages: String
+  ) {
+    /** class loader to load the compile classes */
+    val classLoader = URLClassLoader(arrayOf(outputDirectory.toURI().toURL()),
```
[the remark this answers is not in the document]
Comment (bnorm) [reply core/src/main/kotlin/com/tschuchort/compiletesting/KotlinJsCompilation.kt:152]: Doubtful, I'll remove it.

File: core/src/test/kotlin/com/tschuchort/compiletesting/CompilerPluginsTest.kt [hunk 3]
```diff
@@ -66,6 +66,20 @@ class CompilerPluginsTest {
...
         verify(mockPlugin, atLeastOnce()).registerProjectComponents(any(), any())
 
-        Assertions.assertThat(result.exitCode).isEqualTo(KotlinCompilation.ExitCode.OK)
+        Assertions.assertThat(result.exitCode).isEqualTo(ExitCode.OK)
     }
-}
\ No newline at end of file
+
+    @Test
+    fun `when JS compiler plugins are added they get executed`() {
```
[the remark this answers is not in the document]
Comment (bnorm) [reply core/src/test/kotlin/com/tschuchort/compiletesting/CompilerPluginsTest.kt:73]: This comes back to how are we going to verify the correct JS files? I can write some really basic tests, maybe simple text parsing will be sufficient initially.

**Review by bnorm — COMMENTED** [review]
> I'm hesitant to add any functionality for this to the library right now if it's not straight forward. It's probably best to leave it in an "incomplete" state, let users figure out the best way to inspect generated JS files and wait for their feedback.

This was my feeling as well. However it also means it's going to be difficult to properly unit test as there is no standard way to test the output of the JS files.

> I think the best way is probably to integrate the new functionality into the existing KotlinCompilation class and add an option to enable JS compilation. Otherwise a lot of stuff has to be factored out so it can be shared by the two KotlinCompilation classes, but that would be quite hard to do for all the logic.

I think there's enough difference between the 2 that it would be better to keep separate. For example, JS doesn't have annotation processing, so what happens to all of the kapt options? Let me see if I can create an AbstractKotlinCompilation class for some shared logic.

File: core/src/main/kotlin/com/tschuchort/compiletesting/KotlinJsCompilation.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (tschuchortdev) [reply core/src/main/kotlin/com/tschuchort/compiletesting/KotlinJsCompilation.kt:184]: I don't want to limit the available options, I was just wondering if they should also be added to the Java KotlinCompilation. I usually try to use the same naming as in the `K2JVMCompilerArguments` class. Originally I did it this way instead of letting the user pass in an instance of `K2JVMCompilerArguments` because I wanted to hide all the internals and dependency on kotlin-compiler-embeddable from the user, but I think this is no longer possible because kotlin-compiler-embeddable also declares the `ComponentRegistrar` interface, which users need for compiler plugins. Giving the user full access to `K2JVMCompilerArguments` is probably not a good idea since I am changing the options and it could be confusing.

File: core/src/test/kotlin/com/tschuchort/compiletesting/CompilerPluginsTest.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (tschuchortdev) [reply core/src/test/kotlin/com/tschuchort/compiletesting/CompilerPluginsTest.kt:73]: Something basic should be enough for now. Like testing that the expected JS file is created, testing that the text contains a string `"class Foo"` and `"function bar()"`. We don't really have to test that the right code is generated since we're not testing the compiler itself, but a basic test will hopefully catch regressions and edge cases on different platforms or when compiler internals change.

**Comment by tschuchortdev** [comment]
> I think there's enough difference between the 2 that it would be better to keep separate. For example, JS doesn't have annotation processing, so what happens to all of the kapt options? Let me see if I can create an AbstractKotlinCompilation class for some shared logic.

There's also _a lot_ of duplicated code though and not only trivial things like common compiler arguments. Why not add this functionality to the other class and add a parameter to chose the compilation target (JVM, JS, JVM_JS). If JS is chosen, then the JVM steps can just be skipped.

**Comment by bnorm** [comment]
I think it goes beyond just skipping compile steps though. Consider all the arguments that are allowed to be set from the user perspective. How are we going to make it clear to the end user which options are JVM or JS specific? What about options that are present in some form on both compilers but unique (eg, enabling IR)?

Here's an example of something I would like to avoid:
```kotlin
KotlinCompilation().apply {
    target = KotlinPlatform.JS // <-- compile JS
    useIR = true // <-- JVM-only argument, should be irProduceJs
    ...
}.compile()
```

Given the way the library is currently designed I'm not sure there is a way to avoid this problem without API breaking changes or creating an abstraction over the compiler options.

I agree, there is a lot of duplicated code. I'm currently working to create a base `AbstractKotlinCompilation` that can be shared by both. But ultimately this is your library and if you want to have a single `KotlinCompilation` class say the word and I'll move in that direction.

**Comment by bnorm** [comment]
Still need to add unit tests for JS compilation but I've added the `AbstractKotlinCompilation` base class.

**Comment by tschuchortdev** [comment]
>  Consider all the arguments that are allowed to be set from the user perspective. How are we going to make it clear to the end user which options are JVM or JS specific? What about options that are present in some form on both compilers but unique (eg, enabling IR)?

I see what you mean. It's definitely not ideal. How does kotlinc deal with this problem? IIRC the property names here are the same as the cli arguments for kotlinc, so I think it wouldn't be unreasonable to expect that people who want to change a property also know what it means. `useIR` is confusing though.

**Review by tschuchortdev — COMMENTED** [review]
In general I'm not a fan of base classes for sharing functionality. It is usually a bad idea, especially if the base class expects subclasses to override something. That said, the way you did it is fine and it is probably the path of least resistance forward until a larger rewrite with breaking API changes that can make the whole thing more modular. Unless I can think of something else that could be improved over the weekend I would say this is mostly ready to be merged.

File: core/src/main/kotlin/com/tschuchort/compiletesting/KotlinJsCompilation.kt [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    args.irProduceJs = irProduceJs
+    args.irDce = irDce
+    args.irDceDriven = irDceDriven
+    args.irDcePrintReachabilityInfo = irDcePrintReachabilityInfo
+    args.irOnly = irOnly
+    args.irModuleName = irModuleName
+  }
+
+  /** Runs the compilation task */
+  fun compile(): Result {
```
[the remark this answers is not in the document]
Comment (bnorm) [reply core/src/main/kotlin/com/tschuchort/compiletesting/KotlinJsCompilation.kt:93]: There is a `Result` nested class for both KotlinJsCompilcation and KotlinCompilation. So the return type is different for each which is why it's a little hard to share. Ideally I would have pulled the Result class up to a top level class so it could be shared but that would break compatibility. I wasn't sure the best way to handle Result and ExitCode since they are both nested classes. Should I duplicate them both?

File: core/src/main/kotlin/com/tschuchort/compiletesting/KotlinJsCompilation.kt [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    Please make sure bin/idea.properties is present in the installation directory"
+    See: https://github.com/arturbosch/detekt/issues/630
+    */
+    withSystemProperty("idea.use.native.fs.for.win", "false") {
+      // step 1: compile Kotlin files
+      return makeResult(compileKotlin(sourceFiles, K2JSCompiler(), commonK2JSArgs()))
+    }
+  }
+
+  private fun makeResult(exitCode: KotlinCompilation.ExitCode): Result {
```
[the remark this answers is not in the document]
Comment (bnorm) [reply core/src/main/kotlin/com/tschuchort/compiletesting/KotlinJsCompilation.kt:121]: Originally I pulled ExitCode out as a separate class but then realized that breaks compatibility so put it back. I wanted to share this Enum between compilations since there really isn't a reason for separate ones. But I understand it is a little awkward. Should I duplicate it?

**Review by bnorm — COMMENTED** [review]
`kotlinc` doesn't compile Kotlin/JS, that's the responsibility of [`kotlinc-js`](https://kotlinlang.org/docs/tutorials/javascript/getting-started-command-line/command-line-library-js.html). In fact, if you print the help for `kotlinc`, the usage says `kotlinc-jvm`:

```
% kotlinc -h  
Usage: kotlinc-jvm <options> <source files>
```

This indicates to me at least that the compilers are going to stay separate.

As for the base class, if you have another design in mind I would love to hear it. Delegation over inheritance would be preferred, but I'm not sure how to delegate all the compiler options concisely. I could have only the arguments in the base class and the functionality in something internal to the library?

File: core/src/main/kotlin/com/tschuchort/compiletesting/KotlinJsCompilation.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (tschuchortdev) [reply core/src/main/kotlin/com/tschuchort/compiletesting/KotlinJsCompilation.kt:93]: In that case it's fine the way you did it.

**Comment by tschuchortdev** [comment]
The tests look good to me. Can you add the doc comment, rename the `common` functions and rebase, so I can merge it? Thanks for implementing this feature, I bet a lot of people will appreciate it.

**Comment by tschuchortdev** [comment]
Thanks for your contribution!
